File: PokerGame/PokerHandRank.py
from collections import Counter

class PokerHandRank:

    def __init__(self, hand):
        # Sort the cards by value, then by suit
        self.cards = sorted(hand.cards, key=lambda c: (c.get_value(), c.get_suit_rank()))
        # Determine and store the rank of the hand
        self.rank_value = self.evaluate_hand()


    
    def evaluate_hand(self):
        # Check for each of poker hand in order from strongest to weakest
        if self.is_royal_flush(): return 1
        if self.is_straight_flush(): return 2
        if self.is_four_of_a_kind(): return 3
        if self.is_full_house():    return 4
        if self.is_flush():      return 5
        if self.is_straight():  return 6
        if self.is_three_of_a_kind(): return 7
        if self.is_two_pair(): return 8
        if self.is_pair(): return 9
        return 10
    

    def get_rank_name(self):
        # Return the name for the hand's rank
        if self.is_royal_flush(): return "Royal Flush"
        if self.is_straight_flush(): return "Straight Flush"
        if self.is_four_of_a_kind(): return "Four of a Kind"
        if self.is_full_house(): return "Full House"
        if self.is_flush(): return "Flush"
        if self.is_straight(): return "Straight"
        if self.is_three_of_a_kind(): return "Three of a Kind"
        if self.is_two_pair(): return "Two Pair"
        if self.is_pair(): return "One Pair"
        return "High Card"


    def is_flush(self):
         # Check if all cards have the same suit
        suit = self.cards[0].get_suit()
        return all(card.get_suit() == suit for card in self.cards)
    

    def is_straight(self):
        # Check if all cards form a sequence in value
        values = [card.get_value() for card in self.cards]

        if values == [2,3,4,5,14]:
            return true
        return all(values[i+1] == values[i] + 1 for i in range(len(values) -1))
    

    def is_royal_flush(self):
         # Check for a royal flush: a straight flush starting at 10
        return self.is_straight_flush() and self.cards[0].get_value() == 10
    
        
    def is_straight_flush(self):
         # Check if the hand is both a straight and a flush
        return self.is_flush() and self.is_straight()
    


    def get_value_counts(self):
        # Return a dictionary counting how many times each card value appears
        return Counter(card.get_value() for card in self.cards)
    

    def is_four_of_a_kind(self):
         # Check if any card value appears 4 times
        return 4 in self.get_value_counts().values()
    

    def is_full_house(self):
        # Check for 3 of a kind and a pair
        counts = self.get_value_counts()
        return 3 in counts.values() and 2 in counts.values()
    

    def is_three_of_a_kind(self):
        # Check for 3 of a kind but not a full house
        counts = self.get_value_counts()
        return 3 in counts.values() and not self.is_full_house()
    

    def is_two_pair(self):
         # Count how many values appear exactly twice
        counts = list(self.get_value_counts().values())
        return counts.count(2) == 2
    
    def is_pair(self):
        # Check for exactly one pair (but not two pair)
        counts = list(self.get_value_counts().values())
        return counts.count(2) == 1 and not self.is_two_pair()
    
    
    def __str__(self):
         # Return a string representation of the hand's rank and cards
      #  card_strs = ' '.join(str(card) for card in self.cards)
        return "Jonathan"

File: PokerGame/PokerHandRank.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace PokerGame
{
    public class PokerHandRank
    {
        private int rankValue;
        private List<Card> sortedCards;

        public PokerHandRank(Hand hand)
        {
            // Make a copy of the hand's cards and sort them
            sortedCards = new List<Card>(hand.GetCards());
            hand.Sort();
            rankValue = EvaluateHand();
        }

        public int GetRankValue()
        {
            return rankValue;
        }

        public string GetRankName()
        {
            if (IsRoyalFlush()) return "Royal Flush";
            if (IsStraightFlush()) return "Straight Flush";
            if (IsFourOfAKind()) return "Four of a Kind";
            if (IsFullHouse()) return "Full House";
            if (IsFlush()) return "Flush";
            if (IsStraight()) return "Straight";
            if (IsThreeOfAKind()) return "Three of a Kind";
            if (IsTwoPair()) return "Two Pair";
            if (IsPair()) return "One Pair";
            return "High Card";
        }

        private int EvaluateHand()
        {
            if (IsRoyalFlush()) return 1;
            if (IsStraightFlush()) return 2;
            if (IsFourOfAKind()) return 3;
            if (IsFullHouse()) return 4;
            if (IsFlush()) return 5;
            if (IsStraight()) return 6;
            if (IsThreeOfAKind()) return 7;
            if (IsTwoPair()) return 8;
            if (IsPair()) return 9;
            return 10;
        }

        private bool IsRoyalFlush()
        {
            return IsStraightFlush() && sortedCards[0].GetValue() == 10;
        }

        private bool IsStraightFlush()
        {
            return IsFlush() && IsStraight();
        }

        private bool IsFlush()
        {
            char suit = sortedCards[0].GetSuit();
            foreach (Card card in sortedCards)
            {
                if (card.GetSuit() != suit) return false;
            }
            return true;
        }

        private bool IsStraight()
        {
            for (int i = 0; i < sortedCards.Count - 1; i++)
            {
                if (sortedCards[i + 1].GetValue() != sortedCards[i].GetValue() + 1)
                {
                    return false;
                }
            }
            return true;
        }

        private Dictionary<int, int> GetValueCounts()
        {
            Dictionary<int, int> counts = new Dictionary<int, int>();
            foreach (Card card in sortedCards)
            {
                if (counts.ContainsKey(card.GetValue()))
                {
                    counts[card.GetValue()]++;
                }
                else
                {
                    counts[card.GetValue()] = 1;
                }
            }
            return counts;
        }

        private bool IsFourOfAKind()
        {
            var counts = GetValueCounts();
            return counts.ContainsValue(4);
        }

        private bool IsFullHouse()
        {
            var counts = GetValueCounts();
            return counts.ContainsValue(3) && counts.ContainsValue(2);
        }

        private bool IsThreeOfAKind()
        {
            var counts = GetValueCounts();
            return counts.ContainsValue(3) && !IsFullHouse();
        }

        private bool IsTwoPair()
        {
            var counts = GetValueCounts();
            int pairCount = 0;
            foreach (var count in counts.Values)
            {
                if (count == 2)
                {
                    pairCount++;
                }
            }
            return pairCount == 2;
        }

        private bool IsPair()
        {
            var counts = GetValueCounts();
            return counts.ContainsValue(2) && !IsTwoPair();
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("Rank: ").Append(rankValue).Append(", Cards: ");
            foreach (Card card in sortedCards)
            {
                sb.Append(card.ToString()).Append(" ");
            }
            return sb.ToString().Trim();
        }
    }
}
